File: frontend/src/api/queryDashboard.ts
import { api } from './client';

// ── Types ────────────────────────────────────────────────────

export interface QueryRecord {
  query_id: number;
  query_text: string;
  segment: string | null;
  geo: string | null;
  country: string | null;
  language: string | null;
  source: string;
  job_id: number;
  status: string;
  domains_found: number;
  targets_found: number;
  effectiveness_score: number | null;
  estimated_cost_usd: number;
  is_saturated: boolean;
  created_at: string | null;
}

export interface QueryListResponse {
  total: number;
  page: number;
  page_size: number;
  data: QueryRecord[];
}

export interface SegmentSaturation {
  key: string;
  total: number;
  saturated: number;
  saturation_rate: number;
  total_domains: number;
  total_targets: number;
}

export interface QuerySummaryResponse {
  total_queries: number;
  done_queries: number;
  failed_queries: number;
  total_domains: number;
  total_targets: number;
  total_cost_usd: number;
  saturation_rate: number;
  avg_effectiveness: number | null;
  by_segment: SegmentSaturation[];
  by_geo: SegmentSaturation[];
  by_country: SegmentSaturation[];
  by_source: SegmentSaturation[];
}

export interface FilterOptionsResponse {
  segments: string[];
  geos: string[];
  countries: string[];
  languages: string[];
  sources: string[];
}

export interface GeoEntry {
  key: string;
  country_en: string;
  cities_en: string[];
}

export interface CountryGroup {
  country: string;
  geos: GeoEntry[];
}

export interface GeoHierarchyResponse {
  countries: CountryGroup[];
}

export interface QueryDashboardFilters {
  project_id: number;
  q?: string;
  segment?: string;
  geo?: string;
  country?: string;
  language?: string;
  source?: string;
  status?: string;
  domains_min?: number;
  domains_max?: number;
  targets_min?: number;
  targets_max?: number;
  date_from?: string;
  date_to?: string;
  is_saturated?: boolean;
  sort_by?: string;
  sort_order?: string;
  page?: number;
  page_size?: number;
}

// ── API methods ──────────────────────────────────────────────

export const queryDashboardApi = {
  async listQueries(filters: QueryDashboardFilters): Promise<QueryListResponse> {
    const params = new URLSearchParams();
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        params.append(key, String(value));
      }
    });
    const response = await api.get(`/dashboard/queries?${params.toString()}`);
    return response.data;
  },

  async getSummary(filters: Omit<QueryDashboardFilters, 'sort_by' | 'sort_order' | 'page' | 'page_size'>): Promise<QuerySummaryResponse> {
    const params = new URLSearchParams();
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        params.append(key, String(value));
      }
    });
    const response = await api.get(`/dashboard/queries/summary?${params.toString()}`);
    return response.data;
  },

  async getFilterOptions(projectId: number): Promise<FilterOptionsResponse> {
    const response = await api.get(`/dashboard/queries/filter-options?project_id=${projectId}`);
    return response.data;
  },

  async getGeoHierarchy(): Promise<GeoHierarchyResponse> {
    const response = await api.get('/dashboard/queries/geo-hierarchy');
    return response.data;
  },
};

File: backend/app/schemas/query_dashboard.py
"""Pydantic schemas for the Query Dashboard."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


# ── Single query record ──────────────────────────────────────
class QueryRecord(BaseModel):
    query_id: int
    query_text: str
    segment: Optional[str] = None
    geo: Optional[str] = None
    country: Optional[str] = None
    language: Optional[str] = None
    source: str  # search engine name
    job_id: int
    status: str
    domains_found: int = 0
    targets_found: int = 0
    effectiveness_score: Optional[float] = None
    estimated_cost_usd: float = 0.0
    is_saturated: bool = False
    created_at: Optional[datetime] = None


class QueryListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    data: list[QueryRecord]


# ── Saturation breakdown ─────────────────────────────────────
class SegmentSaturation(BaseModel):
    key: str
    total: int = 0
    saturated: int = 0
    saturation_rate: float = 0.0
    total_domains: int = 0
    total_targets: int = 0


# ── Summary / aggregates ─────────────────────────────────────
class QuerySummaryResponse(BaseModel):
    total_queries: int = 0
    done_queries: int = 0
    failed_queries: int = 0
    total_domains: int = 0
    total_targets: int = 0
    total_cost_usd: float = 0.0
    saturation_rate: float = 0.0
    avg_effectiveness: Optional[float] = None
    by_segment: list[SegmentSaturation] = Field(default_factory=list)
    by_geo: list[SegmentSaturation] = Field(default_factory=list)
    by_country: list[SegmentSaturation] = Field(default_factory=list)
    by_source: list[SegmentSaturation] = Field(default_factory=list)


# ── Filter options (distinct values) ─────────────────────────
class FilterOptionsResponse(BaseModel):
    segments: list[str] = Field(default_factory=list)
    geos: list[str] = Field(default_factory=list)
    countries: list[str] = Field(default_factory=list)
    languages: list[str] = Field(default_factory=list)
    sources: list[str] = Field(default_factory=list)


# ── Geo hierarchy ────────────────────────────────────────────
class GeoEntry(BaseModel):
    key: str
    country_en: str = ""
    cities_en: list[str] = Field(default_factory=list)


class CountryGroup(BaseModel):
    country: str
    geos: list[GeoEntry]


class GeoHierarchyResponse(BaseModel):
    countries: list[CountryGroup]
